### PR TITLE
On macro run error: show status, stdout and stderr

### LINKED ISSUE #4577: ECR.embed with missing file gives an unclear error message
If you have a call to ECR.embed, but the file doesn't exist, you get a compiler error message with an empty "Got:" section and the error message doesn't say anything about a missing file.  I initially hit this with a kemal app, but I have a smaller standalone reproducer below.

`Crystal 0.22.0 (2017-04-20) LLVM 4.0.0`

**foo.cr**

```crystal
require "ecr"
io = STDOUT
ECR.embed("index.ecr", "io")
```

Trying to compile this without the index.ecr present gives:

```text
$ crystal build foo.cr
Error in foo.cr:3: expanding macro

ECR.embed("index.ecr", "io")
^

in macro 'embed' /usr/local/Cellar/crystal-lang/0.22.0/src/ecr/macros.cr:69, line 1:

>  1.     {{ run("ecr/process", "index.ecr", "io") }}
   2.

expanding macro
in foo.cr:3: expanding macro

ECR.embed("index.ecr", "io")
^

in macro 'embed' /usr/local/Cellar/crystal-lang/0.22.0/src/ecr/macros.cr:69, line 1:

>  1.     {{ run("ecr/process", "index.ecr", "io") }}
   2.

Error executing run: ecr/process "index.ecr" "io"

Got:


```

I did not expect the "Got:" part to be empty, and instead expected something closer to the kind of error from a missing require file.

### PR DESCRIPTION
**Pull request by bew**
First PR to display more information about a failing macro run.
Another PR might come to display the errors without the backtrace from `ecr/processor` to properly resolve #4577 

Given thoses files:
```cr
# processor.cr

def func
  raise "Some exception happened" # This will always fail
end

puts %(puts "some stuff on stdout")

begin
  func
rescue e
  STDERR.puts "ERROR: #{e}"
  exit 1
end
```
```cr
# test_run.cr

{{ run "./processor" }}
```
Compilation will give:
```
$ crystal ./test_run.cr
Error in test_run.cr:1: expanding macro

{{ run "./processor" }}
^

in test_run.cr:1: Error executing run (exit code: 1): ./processor 

---- Got on stdout: ----

puts "some stuff on stdout"


---- Got on stderr: ----

ERROR: Some exception happened



{{ run "./processor" }}
   ^~~
```

If you have any other ideas how to present stdout & stderr, I'm open to suggestion.

There is currently no specs on the macro method `run`, should I try to make one for this?

### CODE REVIEW

**Comment by RX14** [comment]
Somehow I don't like the "---- Got on stdout: ----" lines, they seem to have a different look and feel to the rest of the compiler error output. I get that that's a wishy-washy complaint but it feels wrong.

**Comment by bew** [comment]
@RX14 I agree completely, but I don't have any idea how this could be presented.
If you have one, I'll be happy to hear it!

**Comment by straight-shoota** [comment]
Maybe indent the stream output?

```
in test_run.cr:1: Error executing run (exit code: 1): ./processor 

    puts "some stuff on stdout"

stderr:

    ERROR: Some exception happened

    {{ run "./processor" }}
       ^~~
```

**Comment by bew** [comment]
Thanks a lot for your approvals ;)
@straight-shoota I'll do the indent, it's wayy nicer, thanks!

**Comment by bew** [comment]
Following @straight-shoota suggestion, it now looks like:
```
Error in test_run.cr:1: expanding macro

{{ run "./processor" }}
^

in test_run.cr:1: Error executing run (exit code: 1): ./processor 

stdout:

    puts "some stuff on stdout"

stderr:

    ERROR: Some exception happened


{{ run "./processor" }}
   ^~~

```

**Comment by bew** [comment]
Also, everything is bold, do I try to remove it for stdout and stderr?
macro_run_errors

**Comment by RX14** [comment]
@bew `stdout:` and `stderr:` should be bold but the output shouldn't be.

**Comment by bew** [comment]
> @bew `stdout:` and `stderr:` should be bold but the output shouldn't be.

How should I go about this? I couldn't get `colorize` to work to remove any coloring/format of a specific text.

**Comment by RX14** [comment]
@bew In this case you'll probably have to manually insert the "reset" escape code (`"\e[0"`) to reset the formatting, then use colorize after that.

**Comment by bew** [comment]
Rebased. Thanks @RX14 for the tip. Even if I don't like hardcoding the terminal escape code, and kind of bypassing the Colorize module, it has the merit to work!

macro_run_errors

File: src/compiler/crystal/macros/methods.cr [hunk 1]
```diff
@@ -211,11 +211,46 @@ module Crystal
...
-        node.raise "Error executing run: #{original_filename} #{run_args.map(&.inspect).join " "}\n\nGot:\n\n#{result}\n"
+        command = "#{original_filename} #{run_args.map(&.inspect).join " "}"
+
+        message = IO::Memory.new
+        message << "Error executing run (exit code: #{result.status.exit_code}): #{command}\n"
+
+        if result.stdout.empty? && result.stderr.empty?
+          message << "\nGot no output."
+        else
+          message << "\e[0m" if Colorize.enabled?
```
Comment (ysbaddaden) [inline src/compiler/crystal/macros/methods.cr]: Maybe a `Colorize#reset` method would be interesting?

Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: It would do it, yes, should I open another PR for this?

Comment (ysbaddaden) [inline src/compiler/crystal/macros/methods.cr]: Would you mind squashing your commits then add another one that implements `Colorize#reset`?

Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: @ysbaddaden how would it be used?
```cr
message << Colorize.reset
message << "".colorize.reset
# --> does it make sense to have: "abc".colorize.reset   ?
```

Comment (ysbaddaden) [inline src/compiler/crystal/macros/methods.cr]: Yes, I believe `"abc".colorize.reset` makes sense: we want to print a string and make sure the ASCII escape codes have been reset.

Thought this is about disabling escape codes, it's still adding escape codes to the string, so we should follow the actual API (as `.bold`, ...)

Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: for `"abc".colorize.reset`, should I save the current formatting, reset, put `"abc"`, then restore the formatting? Or should I go simple for this PR, and just reset and put the text?

Comment (RX14) [inline src/compiler/crystal/macros/methods.cr]: I actually think that `Colorize.reset` is a better API. Using `String#colorize` implies you want to do something to a string, and most importantly that whatever changes you made will be reset afterwards. With `String#colorize.reset` you don't get that, there's no way to restore the old colours at the end of the string, so I don't think that API makes sense.

Comment (ysbaddaden) [inline src/compiler/crystal/macros/methods.cr]: ok

**Comment by bew** [comment]
So I rebased, squashed, and added `Colorize.reset(io)`.
Good to go?

**Comment by RX14** [comment]
Perhaps the `Colorize.reset` commit should come before the other one so it can use it directly?

**Comment by bew** [comment]
@RX14 done.

File: src/compiler/crystal/macros/methods.cr [hunk 2]
```diff
@@ -211,11 +211,46 @@ module Crystal
...
+          end
+
+          unless result.stderr.empty?
+            message.puts
+            message << "stderr:".colorize.mode(:bold)
+            message.puts
+            message.puts
+            result.stderr.each_line do |line|
+              message << "    "
+              message << line
```
Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: I missed the `\n`

File: src/compiler/crystal/macros/methods.cr [hunk 3]
```diff
@@ -211,11 +211,46 @@ module Crystal
...
+          Colorize.reset(message)
+
+          unless result.stdout.empty?
+            message.puts
+            message << "stdout:".colorize.mode(:bold)
+            message.puts
+            message.puts
+            result.stdout.each_line do |line|
+              message << "    "
+              message << line
```
Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: I missed the `\n`

**Pull request merged**